File: tile.py
from PIL import Image
from io import BytesIO
import tile_utilities

#Represents a single mapzen tile
class Tile:
    def __init__(self, coordinates, zoom, init_scale, filetype, size):
        self.coordinates = coordinates
        self.zoom = zoom
        self.init_scale = init_scale
        self.filetype = filetype
        self.size = size
        self.data = []
        self.im = None
        self.done_decoding = False

    def __eq__(self, other):
        if self.coordinates == other.coordinates:
            return True

    #Retrieve and decode tile data
    def decode(self):
        scale = (2 ** (self.zoom - 10)) * self.init_scale
        self.im = tile_utilities.retrieve_image(self.coordinates, self.zoom, self.filetype, self.size)
        for i in range(0, self.im.size[1]):
            for j in range(0, self.im.size[0]):
                self.data.append(scale * tile_utilities.decode_pixel(self.im.getpixel((j, i))))
        self.done_decoding = True

File: server-app/src/main/java/org/java_websocket/server_app/Tile.java
package org.java_websocket.server_app;

import java.awt.Color;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class Tile {
	private ArrayList<Integer> coordinates;
	private int zoom;
	private int initScale;
	private String filetype;
	private int size;
	private ArrayList<Double> data;
	private BufferedImage im;
	private boolean doneDecoding;
	private Utilities utilities;
	
	//Constructor
	public Tile(ArrayList<Integer> coordinates, int zoom, int initScale, String filetype, int size) {
		this.coordinates = coordinates;
		this.zoom = zoom;
		this.initScale = initScale;
		this.filetype = filetype;
		this.size = size;
		this.data = new ArrayList<Double>();
		this.im = null;
		this.doneDecoding = false;
		this.utilities = new Utilities();
	}
	
	@Override
	public boolean equals(Object other) {
		if(other instanceof Tile) {
			if(((Tile) other).coordinates.equals(this.coordinates)) {
				return true;
			}
		}
		return false;
	}
	
	//Retrieve and decode tile data
	public void decode() {
        double scale = Math.pow(2, (this.zoom - 10)) * this.initScale;
        
        try {
        	this.im = utilities.retrieveImage(this.coordinates, this.zoom, this.filetype, this.size);
        
	        for(int i = this.im.getMinY(); i < this.im.getHeight(); i++) {
	            for(int j = this.im.getMinX(); j < this.im.getWidth(); j++) {
	                this.data.add(scale * utilities.decodePixel(new Color(this.im.getRGB(j, i), true)));
	            }
	        }
	        
	        this.setDoneDecoding();
        }
        catch(Exception e) {
        	System.out.println("File could not be found");
        }
	}
	
	//Set tile state as done decoding
	public synchronized void setDoneDecoding() {
		this.doneDecoding = true;
	}
	
	//Check if tile is done decoding
	public synchronized boolean isDoneDecoding() {
		return this.doneDecoding;
	}
	
	//Return the coordinates of this tile
	public ArrayList<Integer> getCoordinates() {
		return this.coordinates;
	}
	
	//Return the decoded data of this tile
	public ArrayList<Double> getData() {
		return this.data;
	}
}
